cinder/tests/unit/volume/test_volume_retype.py

```diff
@@ -146,3 +147,56 @@ def test_multiattach_to_multiattach_retype(self, mock_authorize):
         mock_authorize.assert_has_calls(
             [mock.call(vol_action_policies.RETYPE_POLICY, target_obj=mock.ANY)
              ])
+
+    def test_retype_with_volume_type_resize_limits(self):
+
+        def _create_min_max_size_dict(min_size, max_size):
+            return {volume_types.MIN_SIZE_KEY: min_size,
+                    volume_types.MAX_SIZE_KEY: max_size}
+
+        def _setup_volume_types():
+            spec_dict = _create_min_max_size_dict(2, 4)
+            sized_vol_type_dict = {'name': 'limit_type',
+                                   'extra_specs': spec_dict}
+            db.volume_type_create(self.context, sized_vol_type_dict)
+            self.sized_vol_type = db.volume_type_get_by_name(
+                self.context, sized_vol_type_dict['name'])
+
+            unsized_vol_type_dict = {'name': 'unsized_type', 'extra_specs': {}}
+            db.volume_type_create(context.get_admin_context(),
+                                  unsized_vol_type_dict)
+            self.unsized_vol_type = db.volume_type_get_by_name(
+                self.context, unsized_vol_type_dict['name'])
+
+        _setup_volume_types()
+        volume_1 = tests_utils.create_volume(
+            self.context,
+            host=CONF.host,
+            status='available',
+            volume_type_id=self.default_vol_type.id,
+            size=1)
+        volume_3 = tests_utils.create_volume(
+            self.context,
+            host=CONF.host,
+            status='available',
+            volume_type_id=self.default_vol_type.id,
+            size=3)
+        volume_9 = tests_utils.create_volume(
+            self.context,
+            host=CONF.host,
+            status='available',
+            volume_type_id=self.default_vol_type.id,
+            size=9)
+
+        self.assertRaises(exception.InvalidInput,
+                          self.volume_api.retype,
+                          self.context, volume_1,
+                          'limit_type',
+                          migration_policy='on-demand')
+        self.assertRaises(exception.InvalidInput,
+                          self.volume_api.retype,
+                          self.context, volume_9,
+                          'limit_type',
+                          migration_policy='on-demand')
+        self.volume_api.retype(self.context, volume_3,
+                               'limit_type', migration_policy='on-demand')
```

Comment:
I somehow think there's something missing here. Why don't we check for the volume actually making it to the new volume type? Just because no exception is raised, doesn't mean everything worked as planned - depending on the implementation which can change in the future. The test would then silently ignore the volume type not being changed.

Reply:
Good catch here, but the almost everything is mocked out during unit tests runs.   So the furthest the code gets is calling the scheduler rpc api call, which is itself is benign because the messaging bus is basically a big mock.   So, the best we could do is simple test to make sure that the volume is left in 'retyping' state.  

I was pretty disappointed after looking into this, that cinder itself doesn't really have an end to end unit test for retype, which is somewhat understandable, but not really at the same time.  There is no functional test for retype, and the contents of test_volume_retype, is woefully lacking, and all of the calls in there result in a volume being left in 'retyping' state.   that's it.